src/vs/workbench/contrib/surveys/browser/ces.contribution.ts

```diff
@@ -0,0 +1,137 @@
+import { Disposable } from 'vs/base/common/lifecycle';
+
+const WAIT_TIME_TO_SHOW_SURVEY = 1000 * 60 * 60; // 1 hours
+const MAX_INSTALL_AGE = 1000 * 60 * 60 * 24; // 24 hours
+const REMIND_LATER_DELAY = 1000 * 60 * 60 * 4; // 4 hours
+const SKIP_SURVEY_KEY = 'ces/skipSurvey';
+const REMIND_LATER_DATE_KEY = 'ces/remindLaterDate';
+
+class CESContribution extends Disposable implements IWorkbenchContribution {
+	private promptScheduler!: RunOnceScheduler;
+
+	constructor(
+		@IStorageService private readonly storageService: IStorageService,
+		@INotificationService private readonly notificationService: INotificationService,
+		@ITelemetryService private readonly telemetryService: ITelemetryService,
+		@IOpenerService private readonly openerService: IOpenerService,
+		@IProductService private readonly productService: IProductService,
+		@optional(ITASExperimentService) private readonly tasExperimentService: ITASExperimentService,
+	) {
+		super();
+
+		if (!productService.cesSurveyUrl) {
+			return;
+		}
+
+		const skipSurvey = storageService.get(SKIP_SURVEY_KEY, StorageScope.GLOBAL, '');
+		if (skipSurvey) {
+			return;
+		}
+		this.promptScheduler = this._register(new RunOnceScheduler(async () => {
+			const sendTelemetry = (userReaction: 'accept' | 'remindLater' | 'neverShowAgain') => {
+				/* __GDPR__
+				"cesSurvey:popup" : {
+					"userReaction" : { "classification": "SystemMetaData", "purpose": "FeatureInsight" }
+				}
+				*/
+				this.telemetryService.publicLog('cesSurvey:popup', { userReaction });
+			};
+
+			const message = await this.tasExperimentService?.getTreatment<string>('CESSurveyMessage') ?? nls.localize('cesSurveyQuestion', 'Got a moment to help the VS Code team? Please tell us about your experience with VS Code so far.');
+			const button = await this.tasExperimentService?.getTreatment<string>('CESSurveyButton') ?? nls.localize('giveFeedback', "Give Feedback");
+
+			this.notificationService.prompt(
+				Severity.Info,
+				message,
+				[{
+					label: button,
+					run: () => {
+						sendTelemetry('neverShowAgain');
+						this.telemetryService.getTelemetryInfo().then(info => {
+							this.openerService.open(URI.parse(`${this.productService.cesSurveyUrl}?o=${encodeURIComponent(platform)}&v=${encodeURIComponent(this.productService.version)}&m=${encodeURIComponent(info.machineId)}}`));
+							this._skipSurvey();
+						});
+					}
+				}, {
+					label: nls.localize('remindLater', "Remind Me later"),
+					run: () => {
+						sendTelemetry('remindLater');
+						this.storageService.store(REMIND_LATER_DATE_KEY, new Date().toUTCString(), StorageScope.GLOBAL, StorageTarget.USER);
+						this._prepareSurvey();
+					}
+				}, {
+					label: nls.localize('neverAgain', "Don't Show Again"),
+					run: () => {
+						sendTelemetry('neverShowAgain');
+						this._skipSurvey();
+					}
+				}],
+				{ sticky: true }
+			);
+		}, WAIT_TIME_TO_SHOW_SURVEY));
+
+		this._prepareSurvey();
```

Comment:
This should be awaited.

Reply:
How would I `await` in a `constructor`, @joaomoreno ? This is why I moved the `await` based code chunk into its own function.

Reply:
Oh I misread: I thought it was inside the scheduler. My bad.